scripts/swebench_lsp_toolchain.py

```
@@ -0,0 +1,4207 @@
+        artifact = entry.get("artifact")
+        source_url = entry.get("source_url")
+        digest = entry.get("artifact_sha256")
+        if not all(isinstance(value, str) and value for value in (artifact, source_url, digest)):
+            raise ToolchainError("toolchain artifact entry is incomplete")
+        target = cache_root / artifact
+        target.parent.mkdir(parents=True, exist_ok=True)
+        if target.is_file() and sha256_file(target) == digest:
+            continue
+        temporary = target.with_name(f".{target.name}.download-{os.getpid()}")
+        try:
+            with urllib.request.urlopen(source_url) as response, temporary.open("wb") as sink:
```

Ruff (0.15.21)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)